### PR TITLE
Sending operations concurrently in peer recovery

### PR DESCRIPTION
**Pull request by dnhatn**
Today, we send operations in phase2 of peer recoveries batch by batch sequentially. Normally that's okay as we should have a fairly small of operations in phase 2 due to the file-based threshold. However, if phase1 takes a lot of time and we are actively indexing, then phase2 can have a lot of operations to replay.

With this change, we will send multiple batches concurrently (defaults to 1) to reduce the recovery time.

Relates #58011

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java [hunk 1]
```diff
@@ -119,7 +119,7 @@ public PeerRecoveryTargetService(ThreadPool threadPool, TransportService transpo
             RecoveryCleanFilesRequest::new, new CleanFilesRequestHandler());
         transportService.registerRequestHandler(Actions.PREPARE_TRANSLOG, ThreadPool.Names.GENERIC,
                 RecoveryPrepareForTranslogOperationsRequest::new, new PrepareForTranslogOperationsRequestHandler());
-        transportService.registerRequestHandler(Actions.TRANSLOG_OPS, ThreadPool.Names.GENERIC, RecoveryTranslogOperationsRequest::new,
+        transportService.registerRequestHandler(Actions.TRANSLOG_OPS, ThreadPool.Names.WRITE, RecoveryTranslogOperationsRequest::new,
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java:122]: I think we should use the WRITE threadpool instead of GENERIC for indexing operations.

**Review by DaveCTurner — COMMENTED** [review]
I think this is a little risky; I've seen cases where it appeared that the load from recovery phase 2 already has a performance impact on indexing into other active shards on the target node. It's certainly nice to have the option of more concurrency in phase 2 but I think we should be cautious about doing so by default.

I'll link such a case to this PR.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 2]
```diff
@@ -661,106 +655,122 @@ void phase2(
...
+        private final long mappingVersion;
+        private int lastBatchCount = 0; // used to estimate the count of the subsequent batch.
+        private final AtomicInteger skippedOps = new AtomicInteger();
+        private final AtomicInteger sentOps = new AtomicInteger();
+        private final AtomicLong targetLocalCheckpoint = new AtomicLong(SequenceNumbers.NO_OPS_PERFORMED);
+
+        OperationBatchSender(long startingSeqNo, long endingSeqNo, Translog.Snapshot snapshot, long maxSeenAutoIdTimestamp,
+                             long maxSeqNoOfUpdatesOrDeletes, RetentionLeases retentionLeases, long mappingVersion,
+                             ActionListener<Void> listener) {
+            super(logger, threadPool.getThreadContext(), listener, maxConcurrentFileChunks, List.of(snapshot));
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:709]: I'm not sure I understand how this adds parallelism: If we only have one item here in the chunks list, then the `MultiChunkTransfer` won't do anything for us will it?

**Review by original-brownbear — COMMENTED** [review]
One question on the mechanics of this, maybe I'm missing something here.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:709]: We have a single snapshot, but we can have multiple batches from it. The MultiChunkTransfer can send multiple batches [without waiting](https://github.com/elastic/elasticsearch/blob/7707c3318111b53b310ce782295d049b06a98d71/server/src/main/java/org/elasticsearch/indices/recovery/MultiChunkTransfer.java#L112) for the acknowledges from the receiver.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:709]: :facepalm: right, sorry for the noise :) My memory on this class betrayed me.

**Review by original-brownbear — COMMENTED** [review]
Looks good to me technically :)

I share Davids concern though. Even more so if we start moving the handler for this to `WRITE` which under load could cause additional trouble from more rejections in a case like the one David linked. Maybe we should just default to the current behavior for now and make this optional?

File: server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: I backed out this change in https://github.com/elastic/elasticsearch/pull/58018/commits/d34db4b50754d3676d76995be0b139e6a32ba095. I think we need a separate change for it.

**Comment by dnhatn** [comment]
@DaveCTurner @original-brownbear Thanks for reviewing. I've added a new recovery setting that controls the number of operation chunks sent in parallel. Could you please take another look?

**Comment by original-brownbear** [comment]
@dnhatn will take a look soon, but FYI there's some random CS failure here in `:test:framework:checkstyleMain` :)

File: docs/reference/modules/indices/recovery.asciidoc [hunk 3]
```diff
@@ -48,3 +48,11 @@ sent in parallel for each recovery. Defaults to `2`.
 +
 You can increase the value of this setting when the recovery of a single shard
 is not reaching the traffic limit set by `indices.recovery.max_bytes_per_sec`.
+
+`indices.recovery.max_concurrent_operation_chunks`::
+(<<cluster-update-settings,Dynamic>>, Expert) Number of operation chunk requests
+sent in parallel for each recovery. Defaults to `1`.
++
+Increasing the value of this setting with care as it can cause recoveries to consume
```
Comment (original-brownbear) [inline docs/reference/modules/indices/recovery.asciidoc:56]: Maybe reword this a little to something like (though maybe @DaveCTurner can use his native speaker powers here to help with wording :)):

Increase the value of this setting with care as higher values will cause increased CPU use, which might negatively impact indexing and search activities.

**Review by original-brownbear — APPROVED** [review]
LGTM :)

File: docs/reference/modules/indices/recovery.asciidoc (hunk 3 above)
Comment (dnhatn) [reply docs/reference/modules/indices/recovery.asciidoc:56]: > though maybe @DaveCTurner can use his native speaker powers here to help with wording

++

File: docs/reference/modules/indices/recovery.asciidoc [hunk 4]
```diff
@@ -48,3 +48,11 @@ sent in parallel for each recovery. Defaults to `2`.
 +
 You can increase the value of this setting when the recovery of a single shard
 is not reaching the traffic limit set by `indices.recovery.max_bytes_per_sec`.
+
+`indices.recovery.max_concurrent_operation_chunks`::
```
Comment (DaveCTurner) [inline docs/reference/modules/indices/recovery.asciidoc:52]: Suggest simply `indices.recovery.max_concurrent_operations`; the fact that operations are sent in chunks (i.e. batches) is a bit too much detail for users here IMO.

File: docs/reference/modules/indices/recovery.asciidoc [hunk 5]
```diff
@@ -48,3 +48,11 @@ sent in parallel for each recovery. Defaults to `2`.
...
+(<<cluster-update-settings,Dynamic>>, Expert) Number of operation chunk requests
```
Comment (DaveCTurner) [inline docs/reference/modules/indices/recovery.asciidoc:53]: Similarly:

```suggestion
(<<cluster-update-settings,Dynamic>>, Expert) Number of operations
```

File: docs/reference/modules/indices/recovery.asciidoc (hunk 3 above)
Comment (DaveCTurner) [reply docs/reference/modules/indices/recovery.asciidoc:56]: Sure, how about this?

```
Concurrently replaying operations during recovery can be very
resource-intensive and may interfere with indexing, search, and other
activities in your cluster. Do not increase this setting without carefully
verifying that your cluster has the resources available to handle the extra
load that will result.
```

File: docs/reference/modules/indices/recovery.asciidoc (hunk 4 above)
Comment (dnhatn) [reply docs/reference/modules/indices/recovery.asciidoc]: ++, see https://github.com/elastic/elasticsearch/pull/58018/commits/e4f1fbfed406b007d3cc24b18781ce1b36607120.

File: docs/reference/modules/indices/recovery.asciidoc (hunk 3 above)
Comment (dnhatn) [reply docs/reference/modules/indices/recovery.asciidoc]: Thanks David. Applied in https://github.com/elastic/elasticsearch/pull/58018/commits/e4f1fbfed406b007d3cc24b18781ce1b36607120.

**Review by DaveCTurner — APPROVED** [review]
Docs & naming LGTM; I haven't reviewed the code change in detail but I think Armin did.

**Comment by dnhatn** [comment]
run elasticsearch-ci/2

**Comment by dnhatn** [comment]
@elasticmachine update branch

**Comment by dnhatn** [comment]
@original-brownbear @DaveCTurner Thanks for reviewing.
